### DISCUSSION TITLE
Two `BelongsTo` fields depending on each other

### DISCUSSION THREAD

**Discussion by JerryBels** [laravel/nova-issues#5748] [Q&A] [answered]
Hey people. I have this entity, `ContactProject`, that links Contacts and Projects in a many to many relationship, except the entity hosts lots of other data (including other relationships) which is why it's a full model and not a simple pivot (also I read [here](https://github.com/laravel/nova-issues/issues/5334#issuecomment-1435691794) and [here](https://github.com/laravel/nova-issues/issues/5173#issuecomment-1357614020) for example that you can't use a Pivot with additional relationships).

So, I have my full model now, and Nova resources are working well. Except that when I create a new `ContactProject` in Nova, I'd like not to add one that already exist. So I thought, contacts' list should update to remove those already linked to the selected project, and project list should update to remove those linked to selected contact. It works great... Except that when you select a project, the contacts list is refreshed without any selected value, and when you select a contact, it's the projects list that removes its selection.

So, you can never have both at the same time 😁 

Here is the code I used:

```php
BelongsTo::make('Contact')
    ->dependsOn(
        ['project'],
        function (BelongsTo $field, NovaRequest $request, FormData $formData) {
            $projectId = $formData->get('project') ?? $formData->get('resource:projects');
            // $contactId = $formData->get('contact') ?? $formData->get('resource:contacts'); // Maybe use this to re-set the value somehow?
            $field->relatableQueryUsing(
                function (NovaRequest $request, Builder $query) use ($projectId) {
                    $query->whereDoesntHave('contactProjects', function (Builder $query) use ($projectId) {
                        $query->where('project_id', $projectId);
                    });
                });
        }
    ),
    
BelongsTo::make('Project')
    ->dependsOn(
        ['contact'],
        function (BelongsTo $field, NovaRequest $request, FormData $formData) {
            $contactId = $formData->get('contact') ?? $formData->get('resource:contacts');
            // $fieldValue = $formData->get('project') ?? $formData->get('resource:projects'); // Maybe use this to re-set the value somehow?
            $field->relatableQueryUsing(
                function (NovaRequest $request, Builder $query) use ($contactId) {
                    $query->whereDoesntHave('contactProjects', function (Builder $query) use ($contactId) {
                        $query->where('contact_id', $contactId);
                    });
                });
        }
    ),
```

So, is there a way to set a value after refresh with `dependsOn` or something else I didn't think of? 

Thanks ahead!

**Comment by davidhemphill** [comment, marked as answer]
This would create a circular dependency, no?

**Reply by JerryBels** [reply]
Well, not really because it should update one depending on the other changing. So if you change `contact` but the selected `project` is still allowed, it should remain as the selected option and thus since there is no change, no further update would occur to `contact`. On the other hand if the selected `project` is unallowed, it will change to empty the choice, and that's it.

The real issue, and it could be considered as a feature request I think, is that a depending field should be able to optionally retain its value if still available. That would resolve my issue (which I have resolved in other ways for now) as well as provide an interesting possibility for depending fields.

**Reply by crynobone** [reply]
When the options get updated via dependsTo the value can't remain the same and and triggers another update, this causes circular dependency.

**Reply by JerryBels** [reply]
Ok. That would be a nice thing to add though. Imagine you have a list of banks, and you want to display allowed user types and loan categories depending on the selected bank. Both of them would `dependsOn` the bank field, and should probably keep their value if already entered and still available.

On the other hand, for a country having a list of regions, for example, you would probably want to reset the list even though a value corresponds to what was already selected. 

That's why I think it should be an option on the `dependsOn` itself.

**Reply by crynobone** [reply]
You should use tier dependency. Hide Region until Country is selected and then only show Region and apply dependency logic for Region based on selected Country.

**Reply by JerryBels** [reply]
I see you don't find the idea appealing :) It's alright, we can do without.
